### PR TITLE
add per-cluster fullsync scheduling

### PR DESCRIPTION
**Pull request by bookshelfdave**
example:
{riak_repl, [
  {data_root, "./data/riak_repl/"},
  % clusters foo + bar have difference intervals (in minutes)
  {fullsync_interval, [{"bar", 1}, {"foo", 2}]}
]}

### CODE REVIEW

File: src/riak_repl_util.erl [hunk 1]
```diff
@@ -562,15 +564,51 @@ configure_socket(Transport, Socket) ->
...
+            ok;
+        {ok, [{_,_} | _] = List} ->
+            case proplists:lookup(Cluster, List) of
+                none -> ok;
+                {_, FullsyncIvalMins} ->
+                    start_fullsync_timer(Pid, FullsyncIvalMins),
+                    ok
+            end;
+        {ok, Tuple} when is_tuple(Tuple) ->
+            {_Cluster, FullsyncIvalMins} = Tuple,
```
Comment (bookshelfdave) [inline src/riak_repl_util.erl]: I should check to see if the first value in the tuple == Cluster.

Comment (Vagabond) [inline src/riak_repl_util.erl]: You could just have 2 clauses; {ok, {Cluster, FullsyncIvalMins}} -> ...; and then an {ok, Tuple} when is_tuple() -> ok;

File: src/riak_repl_util.erl [hunk 2]
```diff
@@ -562,15 +564,51 @@ configure_socket(Transport, Socket) ->
...
+
+schedule_cluster_fullsync(Cluster, Pid) ->
+    case application:get_env(riak_repl, fullsync_interval) of
+        {ok, disabled} ->
+            ok;
+        {ok, [{_,_} | _] = List} ->
+            case proplists:lookup(Cluster, List) of
+                none -> ok;
+                {_, FullsyncIvalMins} ->
+                    start_fullsync_timer(Pid, FullsyncIvalMins),
```
Comment (Vagabond) [inline src/riak_repl_util.erl]: Why not use the erlang:send_after like the olf schedule_fullsync does?

Comment (bookshelfdave) [inline src/riak_repl_util.erl]: The call to riak_repl2_fscoordinator is a cast, not a call. In the older repl, start_fullsync was a call. I felt like there were less changes to the existing code by just using gen_server:cast.

File: src/riak_repl_util.erl [hunk 3]
```diff
@@ -562,15 +564,50 @@ configure_socket(Transport, Socket) ->
...
+                    ok
+            end;
+        {ok, {Cluster, FullsyncIvalMins}} ->
+            start_fullsync_timer(Pid, FullsyncIvalMins);
```
Comment (Vagabond) [inline src/riak_repl_util.erl]: It might be nice to log "Fullsync for Cluster scheduled in 2 hours" or something...

Comment (bookshelfdave) [inline src/riak_repl_util.erl]: Ok. Unless you have a strong opinion regarding hours, I'd rather show minutes. We don't have to have any different cases for intervals smaller than 1 hour.

**Comment by buddhisthead** [comment]
Dave,

This is cool. What do you think about having full sync run "automatically" when we detect the rt dirty flag? It could be some minutes later to avoid slamming a cluster. Maybe some kind of maximum frequency. But it could be cool if it ran when it needed to. Just a thought.

**Comment by bookshelfdave** [comment]
Chris -
  I could see that as being an app.config param:
...
{fullsync_on_connect, false},
{fullsync_on_rtdirty, [params...]}
...

I would add it as a feature for 2.0 to make sure we meet our 1.4 deadline.

**Comment by buddhisthead** [comment]
Yep. I concur about delaying it for 2.0.
Just wanted to toss the idea out there. Thanks for discussing.

**Comment by bookshelfdave** [comment]
@buddhisthead I created a 2.0 issue for that feature: https://github.com/basho/riak_repl/issues/196

**Comment by bookshelfdave** [comment]
TODO:
- [x] remove riak_repl_stats:server_fullsyncs() when fullsync finishes in BNW
- [x] add per cluster server_fullsync stat
- [x] a riak_test that uses per cluster server_fullsync stat

**Comment by bookshelfdave** [comment]
see https://github.com/basho/riak_test/pull/221 for a riak_test.

**Comment by buddhisthead** [comment]
Be aware that with the latest changes to dependencies of protocol buffer APIs, this branch needs to be merged into the latest master so that it will compile. Don't just checkout the branch :-)

**Comment by buddhisthead** [comment]
Besides these comments, everything looks great. Tests Pass.

**Comment by bookshelfdave** [comment]
I rebased and pushed this morning so the deps should be fine now.

**Comment by buddhisthead** [comment]
+1 to merge. Thanks, Dave. Nice work.

**Pull request merged**